### PR TITLE
AMDGPU: Return legal addressmode correctly for flat scratch

### PR DESCRIPTION
**Pull request by ruiling**
Depends on #70634 for test changes.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 1]
```diff
@@ -1357,6 +1357,13 @@ bool SITargetLowering::isLegalGlobalAddressingMode(const AddrMode &AM) const {
   return isLegalMUBUFAddressingMode(AM);
 }
 
+bool SITargetLowering::isLegalFlatScratchAddressingMode(
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:1360]: As an alternative, you could change `isLegalFlatAddressingMode` to take `AS` and `Variant` as arguments.

**Review by jayfoad — APPROVED** [review]
LGTM.

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp (hunk 1 above)
Comment (ruiling) [reply llvm/lib/Target/AMDGPU/SIISelLowering.cpp]: Done in the latest commit.

**Review by jayfoad — APPROVED** [review]
LGTM, thanks.

**Pull request merged**